phabricatoremails/render/mailbatch.py

```diff
@@ -0,0 +1,112 @@
+from dataclasses import dataclass
+from typing import Optional, List, Dict
+
+from phabricatoremails.mail import OutgoingEmail
+from phabricatoremails.render.events.common import Recipient
+from phabricatoremails.render.events.phabricator import Revision
+from phabricatoremails.render.events.phabricator_secure import SecureRevision
+from phabricatoremails.render.template import TemplateStore
+
+
+class MailBatch:
+    """Creates several outgoing emails from a single Phabricator event
+
+    Most Phabricator events trigger multiple emails from the same source data.
+    For example, if a reviewer accepts a revision, then the author and all the
+    other reviewers are notified.
+
+    Different recipients may receive different emails from different templates, but
+    they're all parameterized by the same event.
+    """
+
+    def __init__(self, css_inline, template_store: TemplateStore):
+        self._targets = []
+        self._css_inline = css_inline
+        self._template_store = template_store
+
+    def target(self, recipient: Optional[Recipient], template_name: str, **kwargs):
+        if not recipient or recipient.is_actor:
+            return
+
+        kwargs["recipient_timezone"] = recipient.timezone
+        self._targets.append(Target(template_name, recipient.email, kwargs))
+
+    def target_all(self, recipients: List[Recipient], template_name: str, **kwargs):
+        for recipient in recipients:
+            self.target(recipient, template_name, **kwargs)
+
+    def _process(
+        self,
+        subject: str,
+        template_name: str,
+        recipient_address: str,
+        timestamp: int,
+        template_params: Dict,
+    ):
+        template = self._template_store.get(template_name)
+        html_email, text_email = template.render(self._css_inline, template_params)
+        return OutgoingEmail(
+            template_name,
+            subject,
+            recipient_address,
+            timestamp,
+            html_email,
+            text_email,
+        )
+
+    def process(
+        self,
+        revision: Revision,
+        actor_name: str,
+        unique_number: int,
+        timestamp: int,
+        event,
+    ):
+        return [
+            self._process(
+                subject="D{}: {}".format(revision.id, revision.name),
+                template_name="normal/" + target.template_name,
```

Comment:
This would extract the string from the context in which it is used. I'm going to merge this with the [existing ongoing discussion here](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r404936558).